PR 2.3/2 - refactor(user) - 4 user exception handlers (#790)

Pablo2203 opened the pull request:
**SUMMARY**
Refactorize core exception handlers to return standardized ApiErrorResponseDto (https://github.com/IT-Academy-BCN/ita-challenges-backend/pull/1005 )

This PR is the continuation of the PR #1005  and it is also the third part of the refactorization made in the class _UserGlobalExceptionHandler_. You can check the other two parts in the following links: #1006 &  (Pending PR)

**MODIFIED FILES**

- UserGlobalExceptionHandler : refactored 4 handlers (Two of them are not in use).
- UserGlobalExceptionHandlerTest : refactored tests for the 4 handlers.

**DETAILS**
This PR continues the work on standardizing error responses across the Challenge microservice. Building on https://github.com/IT-Academy-BCN/ita-challenges-backend/pull/996, we now refactor the following handlers to produce structured, consistent error responses:

- [x] handleMethodArgumentNotValidException  **(NOT IN USE)**
- [x] handleUnmodifiableSolutionException
- [x] handleInternalServerErrorException **(NOT IN USE)**
- [x] handleUsernameAlreadyExistsException

Each handler now returns an _ApiErrorResponseDto_ that includes:

_Timestamp_ – when the error occurred
_HTTP Status_ – corresponding to the exception type
_Message_ – human-friendly description
_Field-level errors_ – via FieldErrorDto, if applicable

This ensures all validation and exception errors are clear, consistent, and user-friendly, while technical details are safely logged rather than exposed.

**OUT OF SCOPE**
Handlers refactored in [Taiga #788](https://tree.taiga.io/project/luisvi-ita-challenges/task/788) and [Taiga #790](https://tree.taiga.io/project/luisvi-ita-challenges/t/790).

The two handlers _handleMethodArgumentNotValidException_ & _handleInternalServerErrorException_ are not in use. Tests were made for both of them. 

We hace to discuss the future of both methods in case of future implementations or deletion.

**IMPACT**
API error responses are now standardized across the microservice
Existing unit tests updated and verified green
Messages are more user-friendly and secure (exception details are logged, not returned)

**DEFINITION OF DONE**

 4 handlers refactored to return ApiErrorResponseDto
 Unit tests updated and passing


**Author Self-check**

- [x]  I tested my changes locally and verified they work as expected
- [x]  The PR has a clear and focused purpose (single feature/fix/refactor)
- [x]  Title, description, and changelog are correctly filled
- [x]  All automated checks (SonarCloud, CI) pass
- [x]  I responded to all previous review comments
- [x]  No unrelated commits or merge leftovers remain

zohra-b committed:
created APIErrorResponse dto + FieldError dto

zohra-b committed:
created draft : messages.properties file

zohra-b committed:
added messages.propertie values to adminCreateUser and userSolutionRequest dto

zohra-b committed:
quited comments before push

zohra-b committed:
created shouldBuildAPIErrorResponse() test for ApiErrorResponse + updated userControllerTest+ userGlobalExceptionHandlerTest to use the messageSource in the GlobalUserHanlder

zohra-b committed:
mocked MessageSource in usercontrollerTest

Commit:
refactor/UserGlobalExceptionHandler/handleMethodArgumentNotValidException

zohra-b committed:
added APIErrorResponse tests

zohra-b committed:
added fieldErrorDtoTest class

Commit:
refactor/UserGlobalExceptionHandler/handleMethodArgumentNotValidException/2

Commit:
refactor/UserGlobalExceptionHandler/handleMethodArgumentNotValidException/2

Commit:
refactor/UserGlobalExceptionHandler/handleMethodArgumentNotValidException/2

Commit:
refactor/UserGlobalExceptionHandler/handleUnmodifiableSolutionException/3

Commit:
refactor/UserGlobalExceptionHandler/handleUnmodifiableSolutionException/3

Commit:
refactor/UserGlobalExceptionHandler/handleUnmodifiableSolutionException/3

Commit:
refactor/UserGlobalExceptionHandler/handleUsernameAlreadyExistsException/4

zohra-b committed:
fixed AdminCreateUserRequestDtoTest

Commit:
refactor/UserGlobalExceptionHandler/handleMethodArgumentNotValidExceptionTest

AdriaVH committed:
Add NoArgsConstructor to APIErrorResponse

AdriaVH committed:
Add IMPORT NoArgsConstructor to APIErrorResponse

Commit:
refactor/UserGlobalExceptionHandler/handleMethodUserNameAlreadyExistsTest

Commit:
adding to messages.properties Exception Messages

Commit:
refactor/UserGlobalExceptionHandler/handleMethodUserNameAlreadyExistsTest

Commit:
refactor/UserGlobalExceptionHandler/handleMethodUserNameAlreadyExists

Commit:
test/UserGlobalExceptionHandler/testHandleMethodArgumentNotValidException & testHandleInternalServerErrorException

Commit:
test/UserGlobalExceptionHandler/testHandleMethodArgumentNotValidException & testHandleInternalServerErrorException

Commit:
test:UserGlobalExceptionHandlerTest/testHandleMethodArgumentNotValidException & testHandleInternalServerErrorException

Pablo2203 committed:
Merge branch 'develop' into refactor_user/#790_UserGlobalExceptionHandler-using-ApiErrorResponse

rioacosta commented on itachallenge-user/src/main/java/com/itachallenge/user/exception/UserGlobalExceptionHandler.java:
```diff
@@ -62,42 +69,81 @@ public ResponseEntity<String> handleNotFoundException(NotFoundException e) {
+                .timestamp(Instant.now())
+                .status(HttpStatus.CONFLICT.value())
+                .error("There's an existing solution with status 'ENDED'.")
+                .message("There's an existing solution with status 'ENDED'.")
```
Due to recent merges with changes that come from RP #984, the term ENDED no longer reflect any of the actual status available, in this particular case that word needs to be replaced by 'SUBMITTED' .

rioacosta reviewed (changes requested):
Following the branching for the task of harmonizing error responses with a reactive strategy across all the app, I think this PR with the handlers modification is coherent with the rest of the task and the coding is alright, this PR is well documented and besides of the word change I don't have any other suggestion.

Pablo2203 committed:
Change error message from 'ENDED' to 'SUBMITTED'

In order to harmonize the error responses and following Rio's feedback, I changed the error and message text. 'END' to 'SUBMITTED' in handleUnmodifiableSolutionException method.

Pablo2203 committed:
Update exception message in test for UnmodifiableSolutionException

In order to harmonize the error responses and following Rio's feedback, I changed the error and message text. 'END' to 'SUBMITTED' in handleUnmodifiableSolutionException test.

Pablo2203 committed:
Merge branch 'develop' into refactor_user/#790_UserGlobalExceptionHandler-using-ApiErrorResponse
